gradle-failure-reports/src/main/java/com/palantir/gradle/failurereports/CompileFailuresService.java

```diff
@@ -0,0 +1,162 @@
+        return Splitter.on(COMPILE_ERROR_LAST_LINE_PATTERN)
+                .splitToStream(compilerErrorsByTaskPath.get(taskPath).toString())
+                .map(this::maybeGetFailureReport)
+                .filter(Optional::isPresent)
+                .map(Optional::get);
+    }
+
+    public static Provider<CompileFailuresService> getSharedCompileFailuresService(
+            Project project, FailureReportsExtension failureReportsExtension) {
+        return project.getGradle()
+                .getSharedServices()
+                .registerIfAbsent("compileFailuresService", CompileFailuresService.class, spec -> {
+                    spec.getParameters().getOutputFile().set(failureReportsExtension.getFailureReportOutputFile());
+                    spec.getParameters()
+                            .getCompileOutputFile()
+                            .set(failureReportsExtension.getFailureReportCompileOutputFile());
+                    spec.getParameters().getRootDir().set(project.provider(() -> project.getRootDir()));
+                });
+    }
+
+    private void markCompileErrorStarted(String taskPath) {
+        startedCompileErrorsByTaskPath.put(taskPath, true);
+    }
+
+    private void markCompileErrorFinished(String taskPath) {
+        startedCompileErrorsByTaskPath.put(taskPath, false);
+    }
+
+    private boolean isCompileErrorStarted(String taskPath) {
+        return startedCompileErrorsByTaskPath.getOrDefault(taskPath, false);
+    }
+
+    private Optional<FailureReport> maybeGetFailureReport(String multiLineError) {
+        Matcher matcher = COMPILE_ERROR_PATTERN.matcher(multiLineError);
+        if (!matcher.find()) {
+            return Optional.empty();
+        }
+        String sourcePathFromError = matcher.group("sourcePath");
+        String errorMessage = matcher.group("errorMessage");
+        int lineNumber = Integer.parseInt(matcher.group("lineNumber"));
+
+        return Optional.of(FailureReport.builder()
+                .header(extractCompileErrorHeader(sourcePathFromError, lineNumber, errorMessage))
+                .clickableSource(FailureReporterResources.getRelativePathWithLineNumber(
+                        getParameters().getRootDir().get().toPath(), Path.of(sourcePathFromError), lineNumber))
+                .errorMessage(matcher.group())
+                .build());
+    }
+
+    private static String extractCompileErrorHeader(String sourcePath, int lineNumber, String error) {
+        // the relevant compiler error header is the first line of the error message.
+        int errorExplanationIndex = error.indexOf("\n");
+        int maxIndex = errorExplanationIndex < 0 ? error.length() : errorExplanationIndex;
+        return FailureReporterResources.sourceFileWithErrorMessage(
+                FailureReporterResources.getFileName(sourcePath), lineNumber, error.substring(0, maxIndex));
+    }
+
+    @Override
+    public final void close() throws Exception {
+        JunitReporter.reportFailures(
```

Comment:
all compileFailures errors will be reported at the end of the Buildservice's lifecycle

this is why we shouldn't report again the compile failures in: https://github.com/palantir/gradle-failure-reports/pull/37/files#diff-942f46a8f2f0a9ca8d95c33af6e262c747a8c748d8c9aa6c45cd1efaae43194dR70